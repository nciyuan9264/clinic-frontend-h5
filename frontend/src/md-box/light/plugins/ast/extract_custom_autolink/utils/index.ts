export { HyperNodeType, getHyperLinkNodes } from './parser';

export const completeProtocol = (raw: string) => {
  if (raw.match(/https?:\/\//)) {
    return raw;
  }
  return `https://${raw}`;
};
