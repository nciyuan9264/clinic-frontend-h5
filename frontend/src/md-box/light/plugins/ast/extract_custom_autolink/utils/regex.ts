// -> refer @ea/url-recognition -> refer 飞书云文档
const PROTOCOL = 'https?|s?ftp|ftps|nfs|ssh';
const LINK_SUFFIX =
  "\\b([/#?][\\-A-Z0-9@:%_+.~#?&'/=;$,!\\*\\[\\]\\(\\){}^|<>]*[-A-Z0-9&@#/%=~\\(\\)_|]?)?";
const LOCALHOST_REG = /localhost(:[0-9]{2,5})?/;
const IP_ADDRESS_REG =
  // eslint-disable-next-line max-len
  /(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}(2[0-4][0-9]|25[0-5]|1[0-9]{2}|[1-9][0-9]|[0-9])(:[0-9]{2,5})?/;
const IP_REG = new RegExp(
  `((((${PROTOCOL})://)?(${LOCALHOST_REG.source}))|(((${PROTOCOL})://)(${IP_ADDRESS_REG.source})))${LINK_SUFFIX}`,
  'gi',
);
const URL_HOST_BODY = '([-A-Z0-9%_+~#@][\\-A-Z:0-9%_+~#@]{0,256}\\.){1,50}';
const ANY_TOP_DOMAIN = '[a-z\\-]{2,15}';
const TOP_DOMAIN =
  'com|org|net|int|edu|gov|mil|ac|ad|ae|af|ag|ai|al|am|an|ao|aq|ar|as|at|au|aw|ax|az|ba|bb|bd|be|bf|bg|bh|bi|bj|bm|bn|bo|br|bs|bt|bv|bw|by|bz|ca|cc|cf|cg|ch|ci|ck|cl|cm|cn|co|cr|cu|cv|cw|cx|cy|cz|de|dj|dk|dm|do|dz|ec|ee|eg|eh|er|es|et|eu|fi|fj|fk|fm|fo|fr|ga|gb|gd|ge|gf|gh|gi|gl|gm|gn|gp|gr|gs|gt|gu|gw|gy|hk|hm|hn|hr|ht|hu|id|ie|il|im|in|io|iq|ir|is|it|jm|jo|jp|ke|kg|kh|ki|km|kn|kp|kr|kw|ky|kz|la|lb|lc|li|lk|lr|ls|lt|lu|lv|ly|ma|mc|md|me|mf|mg|mh|mk|ml|mm|mn|mo|mp|mq|mr|ms|mt|mu|mv|mw|mx|my|mz|na|nc|ne|nf|ng|ni|nl|no|np|nr|nu|nz|om|pa|pe|pf|pg|ph|pk|pl|pm|pn|pr|ps|pt|pw|py|qa|re|ro|rs|ru|rw|sa|sb|sc|sd|se|sg|sh|si|sj|sk|sl|sm|sn|so|sr|ss|st|su|sv|sx|sy|sz|tc|td|tf|tg|th|tj|tk|tl|tm|tn|to|tp|tr|tt|tv|tw|tz|ua|ug|uk|um|us|uy|uz|va|vc|ve|vg|vi|vn|vu|wf|ws|ye|yt|za|zm|zw|site|top|wtf|xxx|xyz|cloud|engineering|help|one';
const TOP_DOMAIN_LITE = 'com|cn|tk|de|net|org|uk|info|nl|ru|hk';
// 由于顶级域名不全，故更改为带 协议头的url 忽略上面顶级域名，直接匹配
const URL_REG = new RegExp(
  // eslint-disable-next-line max-len
  `(((${PROTOCOL}):\\/\\/${URL_HOST_BODY}${ANY_TOP_DOMAIN})|(${URL_HOST_BODY}(${TOP_DOMAIN_LITE})))(:[0-9]{2,5})?${LINK_SUFFIX}`,
  'gi',
);
// mailto的贪婪匹配在遇到极长的字符串（无空格的英文）会很慢，需要限长。
const EMAIL_URL_BODY = "[\\w.!#$%&'*+-/=?^_\\`{|}~]{1,2000}@[A-Za-z0-9_.-]+\\.";
const LENGTH_LIMITED_EMAIL_REG = new RegExp(
  `^(?!data:)((mailto:${EMAIL_URL_BODY}${ANY_TOP_DOMAIN})|(${EMAIL_URL_BODY}(${TOP_DOMAIN})))\\b`,
  'gi',
);

export const urlRegExp = new RegExp(
  `(${IP_REG.source})|(${URL_REG.source})|(${LENGTH_LIMITED_EMAIL_REG.source})`,
  'gi',
);
