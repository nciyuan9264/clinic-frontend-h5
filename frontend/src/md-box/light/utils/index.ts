export * from './micromark';
export * from './marked';
export * from './renderer';
export * from './common';
export * from './hop';
export * from './plugin';
