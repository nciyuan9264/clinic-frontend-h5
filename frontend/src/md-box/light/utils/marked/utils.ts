/** 源自 markedjs 源码 */

const escapeTest = /[&<>"']/;
const escapeReplace = new RegExp(escapeTest.source, 'g');
const escapeTestNoEncode = /[<>"']|&(?!(#\d{1,7}|#[Xx][a-fA-F0-9]{1,6}|\w+);)/;
const escapeReplaceNoEncode = new RegExp(escapeTestNoEncode.source, 'g');

const escapeReplacements: Record<string, string> = {
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;',
  '"': '&quot;',
  "'": '&#39;',
};

export const escapeMarkdownText = (html: string, encode = false) => {
  const getEscapeReplacement = (ch: string) => escapeReplacements[ch];

  if (encode) {
    if (escapeTest.test(html)) {
      return html.replace(escapeReplace, getEscapeReplacement);
    }
  } else {
    if (escapeTestNoEncode.test(html)) {
      return html.replace(escapeReplaceNoEncode, getEscapeReplacement);
    }
  }
  return html;
};
