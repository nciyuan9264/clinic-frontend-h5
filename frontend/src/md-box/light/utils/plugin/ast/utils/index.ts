export * from './update_ast';
