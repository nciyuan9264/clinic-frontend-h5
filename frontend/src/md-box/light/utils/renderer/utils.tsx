import React, { createElement } from 'react';

import { isRegExp } from 'lodash-es';
import { DOMNode, Element, Text, domToReact } from 'html-react-parser';

import { RenderFunction } from '../../type';

export const isTextType = (node: DOMNode): node is Text => {
  return node.type === 'text';
};

export const isEndlineText = (node: DOMNode): node is Text => {
  return isTextType(node) && /^\n+$/.test(getInnerText(node));
};

/**
 * 此处断言的是Element的子类型：即为A标签的类型，但是Element不可进一步进行类型收窄，因此
 * 返回node is Element
 */
export const isElementType = (
  node: DOMNode,
  ...names: (string | RegExp)[]
): node is Element => {
  if (!('attribs' in node) || typeof node.attribs !== 'object') {
    return false;
  }
  return (
    node.type === 'tag' &&
    (!names?.length ||
      names.some((item) => {
        if (isRegExp(item)) {
          return item.test(node.name);
        } else {
          return item === node.name;
        }
      }))
  );
};

export const getInnerText = (node: DOMNode): string => {
  if (isTextType(node)) {
    return node.data;
  }
  if (isElementType(node)) {
    return node.children.map(getInnerText).join('');
  }
  return '';
};

/**
 * 检测字符或字符串是否包含中文字符
 * https://github.com/anton-bot/contains-chinese/blob/master/index.js#L7
 * @param char 字符或字符串
 * @returns 是否包含中文字符
 */
export const detectIsChinese = (char: string) => {
  const HAN_REGEX =
    // eslint-disable-next-line max-len
    /[\u2E80-\u2E99\u2E9B-\u2EF3\u2F00-\u2FD5\u3005\u3007\u3021-\u3029\u3038-\u303B\u3400-\u4DB5\u4E00-\u9FD5\uF900-\uFA6D\uFA70-\uFAD9]/;
  return HAN_REGEX.test(char);
};

export const renderDom: RenderFunction<false> = (node) => {
  return <>{domToReact([node])}</>;
};

export const renderReactElement = <P extends Record<string, unknown>>(
  tag: string,
  props: P,
  children?: React.ReactNode[],
) => {
  if (props.className === 'markdown-root') {
    return <>{children}</>;
  } else {
    return (
      <>{createElement(tag, props, children?.length ? children : undefined)}</>
    );
  }
};
