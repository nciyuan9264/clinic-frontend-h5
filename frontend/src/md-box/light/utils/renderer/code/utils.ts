import { DOMNode, Element } from 'html-react-parser';

import { isElementType } from '../utils';

const DEFAULT_LANGUAGE = 'plaintext';

const getCodeElement = (node: DOMNode): Element | undefined => {
  if (isElementType(node, 'code')) {
    return node;
  }

  if (isElementType(node, 'pre')) {
    return node.childNodes.find((childNode): childNode is Element =>
      isElementType(childNode, 'code'),
    );
  }
};

export const getLanguageFromElement = (
  node: DOMNode,
  defaultLanguage = DEFAULT_LANGUAGE,
): string => {
  const codeElement = getCodeElement(node);

  if (codeElement) {
    return (
      codeElement.attribs?.class?.match(/language-(.*)/)?.[1] || defaultLanguage
    );
  }

  return defaultLanguage;
};

export const getMetaFromElement = (node: DOMNode, removeBuiltIn = true) => {
  const codeElement = getCodeElement(node);

  const metaString = codeElement?.attribs['data-meta'];

  if (!metaString) {
    return;
  }

  if (!removeBuiltIn) {
    return metaString;
  }

  return (
    metaString
      .split(/\s+/)
      .filter((item) => !item.match(/^__[^_].*/))
      .join(' ') || undefined
  );
};

export const getIndentFromCodeElememt = (node: DOMNode) => {
  const metaString = getMetaFromElement(node, false);

  if (!metaString) {
    return 0;
  }

  const matchResult = metaString.match(
    /(\s|^)__indent=(?<indent>[0-9]+)(\s|$)/,
  );

  if (!matchResult) {
    return 0;
  }

  const indentValue = Number(matchResult.groups?.indent);

  return isNaN(indentValue) ? 0 : indentValue;
};
