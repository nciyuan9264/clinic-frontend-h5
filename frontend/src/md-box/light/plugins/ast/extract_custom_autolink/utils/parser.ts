import emailValidator from 'email-validator';

import { urlRegExp } from './regex';

export enum HyperNodeType {
  text,
  link,
}

export type HyperNode =
  | {
      type: HyperNodeType.text;
      text: string;
    }
  | {
      type: HyperNodeType.link;
      url: string;
    };

export type HyperNodeTitled = HyperNode & {
  title: string;
};

const openAndCloseCharPairs: Record<string, string> = {
  '(': ')',
  '[': ']',
  '{': '}',
  '<': '>',
};
const openChars = Object.keys(openAndCloseCharPairs);
const closeChars = Object.values(openAndCloseCharPairs);
const unreasonableSuffix = ['.', ',', '!', '?', "'", ':', ';', '@', '#'];

/**
 * 分割合理的URL
 * @returns 返回字符串中合理的URL的长度，范围[0, input.length]
 */
const getAReasonableUrl = (input: string) => {
  let index = -1;
  const stack: Array<{
    char: string;
    index: number;
  }> = [];
  for (let i = 0; i < input.length; ++i) {
    const char = input[i];
    if (openChars.includes(char)) {
      stack.push({
        char,
        index: i,
      });
    } else if (closeChars.includes(char)) {
      if (!stack.length) {
        index = i;
        break;
      } else {
        const { char: lastOpenChar } = stack[stack.length - 1];
        if (char === openAndCloseCharPairs[lastOpenChar]) {
          stack.pop();
        } else {
          ({ index } = stack[0]);
          break;
        }
      }
    }
  }
  if (stack.length) {
    ({ index } = stack[0]);
  }
  if (index === -1) {
    index = input.length;
  }

  while (index > 1 && unreasonableSuffix.includes(input[index - 1])) {
    index--;
  }
  return input.slice(0, index);
};

/**
 * 分割字符串中的超链接和常规字符，输出格式化数据 HyperNode[]
 */
export const getHyperLinkNodes = (str: string): HyperNode[] => {
  const nodes: HyperNode[] = [];
  let cur = 0;
  urlRegExp.lastIndex = 0;
  while (true) {
    const result = urlRegExp.exec(str);
    if (!result) {
      break;
    }

    const { index, 0: content } = result;
    const text = str.slice(cur, index);

    // ① 文字部分
    if (text) {
      nodes.push({
        type: HyperNodeType.text,
        text,
      });
    }

    // ② 链接/markdown/email 部分，email 被渲染为文字，code block / inline code 中的链接也当做文字
    if (content) {
      /** 1. 非正常前缀 */
      const abnormalPrefixMatchResult = /^(@|#|\+|-|%)+/.exec(content);
      if (abnormalPrefixMatchResult) {
        nodes.push({
          type: HyperNodeType.text,
          text: abnormalPrefixMatchResult[0],
        });
        cur = result.index + abnormalPrefixMatchResult[0].length;
        urlRegExp.lastIndex = cur;
      } else {
        /** 2. 合理Url作为前缀 */
        const subUrl = getAReasonableUrl(content);

        if (subUrl) {
          if (emailValidator.validate(subUrl)) {
            nodes.push({
              type: HyperNodeType.text,
              text: subUrl,
            });
          } else {
            nodes.push({
              type: HyperNodeType.link,
              url: subUrl,
            });
          }

          cur = result.index + subUrl.length;
          urlRegExp.lastIndex = cur;
        } else {
          const skipStr = content.slice(0, 1);

          nodes.push({
            type: HyperNodeType.text,
            text: skipStr,
          });

          cur = result.index + skipStr.length;
          urlRegExp.lastIndex = cur;
        }
      }
    }
  }
  if (cur !== str.length) {
    nodes.push({ type: HyperNodeType.text, text: str.slice(cur) });
  }
  return nodes;
};
