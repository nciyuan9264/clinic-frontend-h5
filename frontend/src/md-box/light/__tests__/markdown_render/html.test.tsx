import htmlEntity from 'he';
import { render, act } from '@testing-library/react';

import { MdBoxForTesting } from '../utils';

// eslint-disable-next-line max-lines-per-function
describe('html渲染测试', () => {
  test('默认支持的html标签', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={`
<span>span标签</span>
<u>u标签</u>
br标签<br/>
`}
          autoFixSyntax={{ autoFixEnding: false }}
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();
  });

  test('默认不支持的html标签', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={`
<img src="https://domain.com/index.png" />
<input />
<textarea />
<table />
<html />
<body />
<head />
<meta />
`}
          autoFixSyntax={{ autoFixEnding: false }}
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();
  });

  test('对支持的html标签转义', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={String.raw`
\<span>span标签\</span>
\<u>u标签\</u>
br标签\<br/>
`}
          autoFixSyntax={{ autoFixEnding: false }}
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();
  });

  test('危险标签', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={`
<link href="https://domain.com/index.css" />
<script src="https://domain.com/index.js"/>
`}
          autoFixSyntax={{ autoFixEnding: false }}
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();
  });

  describe('设置 enabledHtmlTags 开启标签支持', () => {
    const htmlHybridTagsText = `
这是<u>下划线</u>文本
这是<span style="color:green;">高亮</span>文本
这是<strong>加粗</strong>文本
这是<em>强调</em>文本
这是<div>普通</div>文本
这是<a href="https://test.com">链接</a>文本
这是<my-custom-tag>自定义标签</my-custom-tag>
    `;

    test('设置 enabledHtmlTags = false 时等效于不设置', async () => {
      const rootViewWithoutSetTags = await act(() =>
        render(
          <MdBoxForTesting
            markDown={htmlHybridTagsText}
            autoFixSyntax={{ autoFixEnding: false }}
          />,
        ),
      );

      const rootViewWithSetTagsFalse = await act(() =>
        render(
          <MdBoxForTesting
            markDown={htmlHybridTagsText}
            autoFixSyntax={{ autoFixEnding: false }}
            enabledHtmlTags={false}
          />,
        ),
      );

      expect(rootViewWithoutSetTags.asFragment()).toStrictEqual(
        rootViewWithSetTagsFalse.asFragment(),
      );
    });

    test('启用部分标签，验证启用生效', async () => {
      const rootViewWithoutSetTags = await act(() =>
        render(
          <MdBoxForTesting
            markDown={htmlHybridTagsText}
            autoFixSyntax={{ autoFixEnding: false }}
          />,
        ),
      );

      const rootViewWithSetSomeTags = await act(() =>
        render(
          <MdBoxForTesting
            markDown={htmlHybridTagsText}
            autoFixSyntax={{ autoFixEnding: false }}
            enabledHtmlTags={['strong', 'em', 'my-custom-tag']}
          />,
        ),
      );

      expect(rootViewWithoutSetTags.asFragment()).toMatchDiffSnapshot(
        rootViewWithSetSomeTags.asFragment(),
      );
    });

    test('二次渲染时启用部分标签，验证启用生效', async () => {
      const rootView = await act(() =>
        render(
          <MdBoxForTesting
            markDown={htmlHybridTagsText}
            autoFixSyntax={{ autoFixEnding: false }}
          />,
        ),
      );

      const prevFragment = rootView.asFragment();

      await act(() =>
        rootView.rerender(
          <MdBoxForTesting
            markDown={htmlHybridTagsText}
            autoFixSyntax={{ autoFixEnding: false }}
            enabledHtmlTags={['strong', 'em', 'my-custom-tag']}
          />,
        ),
      );

      expect(prevFragment).toMatchDiffSnapshot(rootView.asFragment());
    });

    test('启用全部标签，验证启用生效', async () => {
      const rootViewWithoutSetTags = await act(() =>
        render(
          <MdBoxForTesting
            markDown={htmlHybridTagsText}
            autoFixSyntax={{ autoFixEnding: false }}
            enabledHtmlTags={['strong', 'em', 'my-custom-tag']}
          />,
        ),
      );

      const rootViewWithSetAllTags = await act(() =>
        render(
          <MdBoxForTesting
            markDown={htmlHybridTagsText}
            autoFixSyntax={{ autoFixEnding: false }}
            enabledHtmlTags
          />,
        ),
      );

      expect(rootViewWithoutSetTags.asFragment()).toMatchDiffSnapshot(
        rootViewWithSetAllTags.asFragment(),
      );
    });
  });

  test('自定义渲染，验证正确渲染', async () => {
    const customRenderHtmlString = `
这是<strong>加粗</strong>文本
这是<my-custom-tag>自定义渲染<span style="color:green;">高亮中的<u>下划线</u></span>内容</my-custom-tag>文本
这是<blockquote>不支持的标签</blockquote>和后缀文本
`;

    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={customRenderHtmlString}
          enabledHtmlTags={['strong', 'my-custom-tag']}
          renderHtml={({ tagName, props, children }) => {
            if (tagName === 'my-custom-tag') {
              return (
                <div {...props} id="custom-rendered-mark-tag">
                  {children}
                </div>
              );
            }
          }}
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();
  });
});

describe('Data 插槽式标签渲染测试', () => {
  const htmlEncode = (value: string) =>
    htmlEntity.encode(value, {
      strict: false,
    });

  const dataSlotString = `行内插槽标签：

prefix<data-inline type="user" value="${htmlEncode(
    htmlEncode(
      JSON.stringify({
        name: 'xiaoming',
        age: 12,
      }),
    ),
  )}" alt="不支持的标签" />suffix

单行插槽标签：

prefix<data-block type="user" value="${htmlEncode(
    htmlEncode(
      JSON.stringify({
        name: 'xiaoming',
        age: 12,
      }),
    ),
  )}" alt="不支持的标签" />suffix

这是<my-custom-tag>未开启的标签</my-custom-tag>

这是<span>已开启的标签</span>
`;

  test('自定义渲染，验证正确渲染', async () => {
    const rootView = await act(() =>
      render(
        <MdBoxForTesting
          markDown={dataSlotString}
          renderDataSlot={({ display, type, value, alt, children }) => {
            const text = `类型：${type}, 兜底内容：${alt}, 值：${JSON.stringify(
              value,
            )}`;

            const Tag = display === 'inline' ? 'span' : 'div';

            return (
              <Tag>
                {text}
                {children}
              </Tag>
            );
          }}
        />,
      ),
    );

    expect(rootView.asFragment()).toMatchSnapshot();
  });
});
