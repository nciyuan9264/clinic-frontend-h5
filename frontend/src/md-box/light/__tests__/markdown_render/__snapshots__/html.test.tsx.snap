// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data 插槽式标签渲染测试 自定义渲染，验证正确渲染 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="false"
    dir="ltr"
    theme-mode="light"
  >
    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      行内插槽标签：
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      prefix
      <span>
        类型：user, 兜底内容：不支持的标签, 值：{"name":"xiaoming","age":12}
      </span>
      suffix
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      单行插槽标签：
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      prefix
      <div>
        类型：user, 兜底内容：不支持的标签, 值：{"name":"xiaoming","age":12}
      </div>
      suffix
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      这是&lt;my-custom-tag&gt;未开启的标签
    </div>
    <br
      class="container wrapper non-select"
    />
    

    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      这是
      <span>
        已开启的标签
      </span>
    </div>
    <br
      class="container wrapper non-select"
    />
    

  </div>
</DocumentFragment>
`;

exports[`html渲染测试 危险标签 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="false"
    dir="ltr"
    theme-mode="light"
  >
    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      &lt;link href="https://domain.com/index.css" /&gt;
      <br
        class="container wrapper non-select"
      />
      &lt;script src="https://domain.com/index.js"/&gt;
    </div>
    <br
      class="container wrapper non-select"
    />
    

  </div>
</DocumentFragment>
`;

exports[`html渲染测试 对支持的html标签转义 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="false"
    dir="ltr"
    theme-mode="light"
  >
    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      &lt;span&gt;span 标签 &lt;/span&gt;
      <br
        class="container wrapper non-select"
      />
      &lt;u&gt;u 标签 &lt;/u&gt;
      <br
        class="container wrapper non-select"
      />
      br 标签 &lt; br/&gt;
    </div>
    <br
      class="container wrapper non-select"
    />
    

  </div>
</DocumentFragment>
`;

exports[`html渲染测试 自定义渲染，验证正确渲染 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="false"
    dir="ltr"
    theme-mode="light"
  >
    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      这是
      <strong>
        加粗
      </strong>
      文本
      <br
        class="container wrapper non-select"
      />
      这是
      <div
        id="custom-rendered-mark-tag"
      >
        自定义渲染
        <span
          style="color: green;"
        >
          高亮中的
          <u>
            下划线
          </u>
        </span>
        内容
      </div>
      文本
      <br
        class="container wrapper non-select"
      />
      这是&lt;blockquote&gt;不支持的标签&lt;/blockquote&gt;和后缀文本
    </div>
    <br
      class="container wrapper non-select"
    />
    

  </div>
</DocumentFragment>
`;

exports[`html渲染测试 设置 enabledHtmlTags 开启标签支持 二次渲染时启用部分标签，验证启用生效 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -24,15 +24,25 @@
        </span>
        文本
        <br
          class="container wrapper non-select"
        />
-       这是&lt;strong&gt;加粗&lt;/strong&gt;文本
+       这是
+       <strong>
+         加粗
+       </strong>
+       文本
        <br
          class="container wrapper non-select"
        />
-       这是&lt;em&gt;强调&lt;/em&gt;文本
+       这是
+       <em
+         style="margin-right: 2px;"
+       >
+         强调
+       </em>
+       文本
        <br
          class="container wrapper non-select"
        />
        这是&lt;div&gt;普通&lt;/div&gt;文本
        <br
@@ -40,11 +50,14 @@
        />
        这是&lt;a href="https://test.com"&gt;链接&lt;/a&gt;文本
        <br
          class="container wrapper non-select"
        />
-       这是&lt;my-custom-tag&gt;自定义标签
+       这是
+       <my-custom-tag>
+         自定义标签
+       </my-custom-tag>
      </div>
      <br
        class="container wrapper non-select"
      />
      "
`;

exports[`html渲染测试 设置 enabledHtmlTags 开启标签支持 启用全部标签，验证启用生效 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -42,23 +42,38 @@
        </em>
        文本
        <br
          class="container wrapper non-select"
        />
-       这是&lt;div&gt;普通&lt;/div&gt;文本
+       这是
+     </div>
      <br
        class="container wrapper non-select"
      />
-       这是&lt;a href="https://test.com"&gt;链接&lt;/a&gt;文本
+     <div>
+       普通
+     </div>
+     文本
      <br
        class="container wrapper non-select"
      />
      这是
-       <my-custom-tag>
-         自定义标签
-       </my-custom-tag>
-     </div>
+     <a
+       class="link"
+       href="https://test.com"
+       target="_blank"
+     >
+       链接
+     </a>
+     文本
+     <br
+       class="container wrapper non-select"
+     />
+     这是自定义标签
+     <div
+       class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
+     />
      <br
        class="container wrapper non-select"
      />
      
"
`;

exports[`html渲染测试 设置 enabledHtmlTags 开启标签支持 启用部分标签，验证启用生效 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -24,15 +24,25 @@
        </span>
        文本
        <br
          class="container wrapper non-select"
        />
-       这是&lt;strong&gt;加粗&lt;/strong&gt;文本
+       这是
+       <strong>
+         加粗
+       </strong>
+       文本
        <br
          class="container wrapper non-select"
        />
-       这是&lt;em&gt;强调&lt;/em&gt;文本
+       这是
+       <em
+         style="margin-right: 2px;"
+       >
+         强调
+       </em>
+       文本
        <br
          class="container wrapper non-select"
        />
        这是&lt;div&gt;普通&lt;/div&gt;文本
        <br
@@ -40,11 +50,14 @@
        />
        这是&lt;a href="https://test.com"&gt;链接&lt;/a&gt;文本
        <br
          class="container wrapper non-select"
        />
-       这是&lt;my-custom-tag&gt;自定义标签
+       这是
+       <my-custom-tag>
+         自定义标签
+       </my-custom-tag>
      </div>
      <br
        class="container wrapper non-select"
      />
      "
`;

exports[`html渲染测试 默认不支持的html标签 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="false"
    dir="ltr"
    theme-mode="light"
  >
    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      &lt;img src="https://domain.com/index.png" /&gt;
      <br
        class="container wrapper non-select"
      />
      &lt;input /&gt;
      <br
        class="container wrapper non-select"
      />
      &lt;textarea /&gt;
      <br
        class="container wrapper non-select"
      />
      &lt;table /&gt;
      <br
        class="container wrapper non-select"
      />
      &lt;html /&gt;
      <br
        class="container wrapper non-select"
      />
      &lt;body /&gt;
      <br
        class="container wrapper non-select"
      />
      &lt;head /&gt;
      <br
        class="container wrapper non-select"
      />
      &lt;meta /&gt;
    </div>
    <br
      class="container wrapper non-select"
    />
    

  </div>
</DocumentFragment>
`;

exports[`html渲染测试 默认支持的html标签 1`] = `
<DocumentFragment>
  <div
    class="container flow-markdown-body theme-default"
    data-show-indicator="false"
    dir="ltr"
    theme-mode="light"
  >
    <div
      class="auto-hide-last-sibling-br paragraph paragraph-element br-paragraph-space"
    >
      <span>
        span 标签
      </span>
      <br
        class="container wrapper non-select"
      />
      <u>
        u 标签
      </u>
      <br
        class="container wrapper non-select"
      />
      br 标签
      <br
        class="container wrapper non-select"
      />
    </div>
    <br
      class="container wrapper non-select"
    />
    

  </div>
</DocumentFragment>
`;
